[2단계 - 로그인 구현하기] 줄리(한승아) 미션 제출합니다.

jbilee opened the pull request:
👋 안녕하세요, 레몬~!

코드에 분기점이 점점 많아져서 그런지 리팩터링이 절실한 결과물이 나왔습니다.. 🥹 3단계를 진행하기 전에 당장 개선이 필요한 부분은 리뷰 반영하면서 수정하고 싶습니다!!

### 이번 단계 고민 포인트
- 지금은 하나의 `RequestHandler`가 HTTP 요청을 GET, POST 등 알맞은 메서드에 매핑하고 처리하는 작업을 전부 수행하고 있습니다. 이 중에서 어떤 기능을 다른 클래스의 책임으로 분리하면 좋을지 고민됩니다.

jbilee committed:
feat: HTTP 요청 메세지에 대한 객체 생성 로직

- HTTP 요청의 메서드, URI, 헤더를 한곳에서 관리하는 클래스 구현

jbilee committed:
refactor: HTTP 응답 메세지 생성 로직 리팩터링

- HTTP 응답 메세지 생성 시 외부에서 상태 및 헤더 값을 입력 받아 상황에 맞는 메세지를 생성하도록 수정

jbilee committed:
refactor: HTTP 요청 처리 로직 분리

- HTTP 요청에 대한 상세 처리 및 응답 메세지 반환 로직은 Http11Processor 외 별도의 클래스에서 수행하도록 수정

jbilee committed:
feat: 302 상태코드 반환 기능

- 로그인 성공 시 302 응답을 반환하고 index 페이지로 리다이렉팅되도록 구현

jbilee committed:
feat: 요청 본문 추출 기능

jbilee committed:
feat: 회원 가입 기능

jbilee committed:
refactor: 로그인 시 /login에 POST 요청으로 보내도록 변경

jbilee committed:
chore: 의존성 스타일 컨벤션 일체화

jbilee committed:
feat: 로그인 및 회원가입 시 UUID 생성 및 쿠키로 반환 기능

jbilee committed:
feat: 사용자 세션 관리 기능

jbilee committed:
feat: 잘못된 로그인에 대한 401 페이지 리다이렉팅 처리

jbilee committed:
refactor: ResponseBuilder 클래스 리팩터링

- 응답 본문이 없어도 Content-Length 헤더를 추가하던 로직 제외
- 유효한 데이터가 있을 경우에만 HTTP 메세지에 추가하도록 개선

jbilee committed:
refactor: GET 요청 처리 로직 리팩터링

- 모든 GET 요청에 대한 일괄 처리가 가능하도록 로직 개선
- 쿼리 파라미터 추출 로직을 프로세서로 이동하여, HttpRequest 객체에 상태로 저장하도록 수정

jbilee committed:
fix: 302 Found 응답 관련 버그 수정

- 응답 헤더 작성 오류로 인해 리다이렉트되지 않았던 이슈 해결

jbilee committed:
refactor: RequestHandler 리팩터링

- POST 요청 처리 로직에 반복되는 코드 함수화
- HTTP 상태코드에 대한 enum 클래스 생성

jbilee committed:
feat: 로그인한 사용자 리다이렉팅 기능

- /login 또는 /register 페이지에 접근했을 때, 현재 로그인 세션 id가 유효할 경우 index 페이지로 리다이렉팅되도록 구현

jbilee committed:
fix: 세션 관련 누락된 예외 상황에 대한 처리 추가

dbsdndcks commented on tomcat/src/main/java/org/apache/catalina/SessionManager.java:
```diff
@@ -0,0 +1,33 @@
+
+    public static SessionManager getInstance() {
+        return INSTANCE;
+    }
```
저도 쓰레드가 여러개인 상황에서 '반복적인 객체 생성이 괜찮은 것인가' 에대해 고민을 했었고 싱글톤으로 구현을 해보았었는데요. 쓰레드 안정성이라는 것을 처음 듣게 되었습니다. 마침 쓰레드에 관한 수업이 있기도 하니 같이 공부해 보면 좋을 것 같습니다!
비슷한 상황에 대해서 정리한 글 공유드릴게요 ! 
[링크] (https://shadowed-plier-7d6.notion.site/269bd3f53d9080d9a859f3cf14df10fd?pvs=73)

jbilee replied:
🙏✨ 소중한 자료 감사합니다!! 멀티스레드 환경에서 race condition이 발생할 수 있는 코드가 여기 외에 다른 곳에도 있어서, 전부 수정해두겠습니다!

jbilee replied:
[refactor: Session 관련 상태의 자료구조 변경](https://github.com/woowacourse/java-http/pull/920/commits/6bf36bcd477488a61e5df4bda188d5a4a39d0ec4)

dbsdndcks commented on tomcat/src/main/java/org/apache/coyote/RequestHandler.java:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    public String handle(final HttpRequest request) throws IOException {
+        if (request.method().equals("GET")) {
+            return handleGet(request);
+        }
+        if (request.method().equals("POST")) {
+            return handlePost(request);
+        }
+        return null;
```
1. handle의 분기를 method로 하는것도 좋지만 "/login", "/register" 와 같은 엔드포인트로 분기를 나누는 것은 어떨까요?? 추후 리팩토링 단계에서 가독성 향상에 도움이 될것 같습니다!.
2. null 반환 보다는 예외 혹은 오류 형식 응답을 보내는 것도 좋을 것 같습니다!

jbilee replied:
1. 피드백 감사합니다! 이 부분은 엔드포인트별로 전용 클래스를 따로 만들면 어떨까 싶습니다 👀
2. 지금 코드에서는 null값으로 응답이 전달될 수 있군요! 이 부분 수정하겠습니다!

jbilee replied:
[refactor: RequestHandler 클래스 리팩터링](https://github.com/woowacourse/java-http/pull/920/commits/f1f888bb8e8c5aec7ac287a319b36c36c8ed5ec1)
[refactor: 지원하지 않는 요청에 대한 핸들링 로직 리팩터링](https://github.com/woowacourse/java-http/pull/920/commits/d75f2623d280dac2bc2ecd683cd1b187d5da8bf7)

dbsdndcks commented on tomcat/src/main/java/org/apache/coyote/ResponseBuilder.java:
```diff
@@ -2,6 +2,7 @@
+import java.util.Map;
 
 public class ResponseBuilder {
 
```
private enum으로 MediaTypes을 지정하셨는데 이유가 궁금해요!

jbilee replied:
현재 구조에서는 응답 메세지를 생성하는 책임을 가진 클래스만 MediaType에 대한 정보를 아는 것이 역할에 부합한다고 판단돼서, 응집도를 높이고 불필요한 전역 상수를 줄이기 위해 ResponseBuilder 안에 두었습니다! 

MediaTypes에 더 복잡한 로직을 추가하게 되거나 MIME 타입이 더 많아지면 클래스로 분리하면 좋을 것 같은데, 동시에 이번 미션에서는 중요 포인트는 아닌 것 같다는 생각도 듭니다. 👀

dbsdndcks commented on tomcat/src/main/java/org/apache/coyote/RequestHandler.java:
```diff
@@ -0,0 +1,123 @@
+            if (request.headers().containsKey("Cookie") && request.headers().get("Cookie").contains("JSESSIONID")) {
+                return handleRedirect(request);
+            }
+            final byte[] body = ResourceLoader.get(uri + ".html");
```
byte 형태로 가져오신 이유가 있을까요? responseBuilder.build 를 보니까 마지막에 new String(body)으로 콘텐츠를 넣어주셨는데 ResourceLoader.get에서 바로 String으로 만들어서 가져와도 될 것 같은데 이유가 궁금합니다!

jbilee replied:
만약에 클라이언트가 요청한 데이터가 텍스트 형식이 아닌 다른 미디어 파일일 경우(.mp3 등) String 대신 순수 byte[]로 반환하는 상황을 고려한 것인데, 정작 응답 body를 생성하는 로직에서는 이 부분을 따로 처리하고 있지 않았군요.. 😭

요청 Content-Type에 따라 String 또는 바이너리로 반환하도록 수정해보려 합니다!

jbilee replied:
클라이언트에 응답하는 OutputStream에 결국 byte[] 타입을 전달해야 해서, 응답 메세지 자체를 String 대신 byte[]로 반환하도록 수정해보았습니다!

[refactor: RequestHandler에서 응답 메세지를 byte로 반환하도록 변경](https://github.com/woowacourse/java-http/pull/920/commits/417a79190f5d4af62fb0b329fb7716274e5ab5ba)

dbsdndcks commented on tomcat/src/main/java/org/apache/coyote/RequestHandler.java:
```diff
@@ -0,0 +1,123 @@
+        return responseBuilder.build(null, HttpStatus.FOUND, null, headers);
+    }
+
+    private String handlePost(final HttpRequest request) {
```
메서드 분리 혹은 위에서 언급 했던 것처럼 api 엔드포인트 별로 클래스 분리가 되면 좋을 것 같습니다! 만약 로그인 , 회원가입 말고 다른 api 기능이 추가 된다면 그때마다 handlePost의 코드 길이가 길어질 것 같아요! 
이 부분은 리팩토링 단계에서 참고 해주시면 될 것 같아요!

jbilee replied:
확인했습니다, 감사합니다! 리팩터링하면서 더 적절한 설계를 고민해보겠습니다 🤓

dbsdndcks commented on tomcat/src/main/java/org/apache/coyote/http11/Http11Processor.java:
```diff
@@ -70,11 +51,48 @@ public void process(final Socket connection) {
+        char[] bodyChars = new char[length];
+        int read = bufferedReader.read(bodyChars, 0, length);
+        bodyBuilder.append(bodyChars, 0, read);
+
```
추후에 메서드 분리 등을 해주시면 가독성이 향상 될 것 같습니다!!

dbsdndcks reviewed (changes requested):
안녕하세요 줄리~ 몇가지 코멘트 남겼습니다! 확인 부탁드리겠습니다 😊

jbilee committed:
refactor: RequestHandler에서 응답 메세지를 byte로 반환하도록 변경

- OutputStream에 byte[] 타입을 전달해야 하는 점을 고려하여, 응답 메세지를 byte[]로 바로 반환해 추가적인 로직 처리가 불필요하도록 개선

jbilee committed:
refactor: 지원하지 않는 요청에 대한 핸들링 로직 리팩터링

dbsdndcks commented on tomcat/src/main/java/com/techcourse/Service.java:
```diff
@@ -10,15 +10,17 @@ public class Service {
 
     private static final Logger log = LoggerFactory.getLogger(Service.class);
```
줄리 ! 이부분을 확인해 봤는데 아이디가 다를때에만 401 페이지로 리다이렉트 되고 있어요 !
4단계 힌트를 보면 checkPassword 메서드가 있는데 구구가 의도한 로그인은 비밀번호 검증도 있지 않을까 싶습니다!
<img width="436" height="187" alt="image" src="https://github.com/user-attachments/assets/09fd228d-5ad3-4130-ba38-4f9027cc0219" />

jbilee replied:
비밀번호 검증이 누락됐군요! 추가하도록 하겠습니다!

jbilee replied:
[fix: 로그인 기능 버그 수정](https://github.com/woowacourse/java-http/pull/920/commits/0824f93a8373217b22787785c3be054cd4d761c9)

dbsdndcks reviewed (changes requested):
줄리 코멘트 단 부분만 확인해주시면 다음번엔 머지하겠습니다!

jbilee committed:
fix: 로그인 기능 버그 수정

- 잘못된 로그인 정보를 입력해도 로그인을 허용했던 로직 수정

jbilee committed:
refactor: RequestHandler 클래스 리팩터링

- 엔드포인트별로 HTTP 요청을 처리할 로직 분리

jbilee committed:
refactor: Session 관련 상태의 자료구조 변경

- 동시성을 고려하여 HashMap에서 ConcurrentHashMap으로 수정

dbsdndcks commented:
고생하셨어요 줄리! 다음 단계도 화이팅입니다!
